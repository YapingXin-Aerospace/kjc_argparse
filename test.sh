#!/bin/bash

script_dir="${BASH_SOURCE%/*}"
prog="$script_dir/full_example"

function run {
	echo "$@"
	$@
}

function run_tests {
	run $prog
	
	run $prog --help
	
	run $prog -h
	
	run $prog --test
	
	run $prog -f
	
	run $prog --flag -f
	
	run $prog -ffffffff
	
	run $prog -o -f
	
	run $prog -of
	
	run $prog -ofo
	
	run $prog -f -oo
	
	run $prog -f -o --once
	
	run $prog --flag=test
	
	run $prog -s foo
	
	run $prog --my-string-argument foo
	
	run $prog --my-string-argument=foo
	
	run $prog -f -i 42
	
	run $prog -f --my-int-argument 42
	
	run $prog -f --my-int-argument=42
	
	run $prog -fi 42
	
	run $prog -if 42
	
	run $prog -f test
	
	run $prog -f -- here are some args
	
	run $prog -tHfots lol --my-int-argument=42 -- test lol omg --help -o -t
	
	run $prog -
	
	run $prog ---
	
	run $prog -f- test
	
	run $prog -=
	
	run $prog -=test
	
	run $prog --=
	
	run $prog --=test
}

run_tests >test_out.actual 2>test_err.actual
diff test_out.{expected,actual} && diff test_err.{expected,actual} && echo "All tests passed!" || echo "Tests failed."
