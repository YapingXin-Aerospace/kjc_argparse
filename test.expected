./full_example
All done with argument parsing!
ERROR --flag is required!
Usage: ./full_example [--fHhiost] [extra args...]
Options:
    -h, --help                Display this help message
        --hello               Say hello!
    -H                        Hello but in caps
    -t, --test                This is a test lol
    -f, --flag                Turns this flag on
    -o, --once                This flag may only be set once
    -i, --my-int-argument     [int] This argument expects an integer value
    -s, --my-string-argument  [string] This argument expects a string value
    --                        All subsequent arguments will not be processed as options
./full_example --help
Usage: ./full_example [--fHhiost] [extra args...]
Options:
    -h, --help                Display this help message
        --hello               Say hello!
    -H                        Hello but in caps
    -t, --test                This is a test lol
    -f, --flag                Turns this flag on
    -o, --once                This flag may only be set once
    -i, --my-int-argument     [int] This argument expects an integer value
    -s, --my-string-argument  [string] This argument expects a string value
    --                        All subsequent arguments will not be processed as options
./full_example -h
Usage: ./full_example [--fHhiost] [extra args...]
Options:
    -h, --help                Display this help message
        --hello               Say hello!
    -H                        Hello but in caps
    -t, --test                This is a test lol
    -f, --flag                Turns this flag on
    -o, --once                This flag may only be set once
    -i, --my-int-argument     [int] This argument expects an integer value
    -s, --my-string-argument  [string] This argument expects a string value
    --                        All subsequent arguments will not be processed as options
./full_example --test
Test: 1 2 3 4 5 6 7 8 9 10
All done with argument parsing!
ERROR --flag is required!
Usage: ./full_example [--fHhiost] [extra args...]
Options:
    -h, --help                Display this help message
        --hello               Say hello!
    -H                        Hello but in caps
    -t, --test                This is a test lol
    -f, --flag                Turns this flag on
    -o, --once                This flag may only be set once
    -i, --my-int-argument     [int] This argument expects an integer value
    -s, --my-string-argument  [string] This argument expects a string value
    --                        All subsequent arguments will not be processed as options
./full_example -f
All done with argument parsing!
./full_example --flag -f
All done with argument parsing!
./full_example -ffffffff
All done with argument parsing!
./full_example -o -f
All done with argument parsing!
./full_example -of
All done with argument parsing!
./full_example -ofo
Flag --once given multiple times!
./full_example -f -oo
Flag --once given multiple times!
./full_example -f -o --once
Flag --once given multiple times!
./full_example --flag=test
Argument "--flag=test" has an embedded value but doesn't expect any value.
./full_example -s foo
--my-string-argument foo
All done with argument parsing!
ERROR --flag is required!
Usage: ./full_example [--fHhiost] [extra args...]
Options:
    -h, --help                Display this help message
        --hello               Say hello!
    -H                        Hello but in caps
    -t, --test                This is a test lol
    -f, --flag                Turns this flag on
    -o, --once                This flag may only be set once
    -i, --my-int-argument     [int] This argument expects an integer value
    -s, --my-string-argument  [string] This argument expects a string value
    --                        All subsequent arguments will not be processed as options
./full_example --my-string-argument foo
--my-string-argument foo
All done with argument parsing!
ERROR --flag is required!
Usage: ./full_example [--fHhiost] [extra args...]
Options:
    -h, --help                Display this help message
        --hello               Say hello!
    -H                        Hello but in caps
    -t, --test                This is a test lol
    -f, --flag                Turns this flag on
    -o, --once                This flag may only be set once
    -i, --my-int-argument     [int] This argument expects an integer value
    -s, --my-string-argument  [string] This argument expects a string value
    --                        All subsequent arguments will not be processed as options
./full_example --my-string-argument=foo
--my-string-argument foo
All done with argument parsing!
ERROR --flag is required!
Usage: ./full_example [--fHhiost] [extra args...]
Options:
    -h, --help                Display this help message
        --hello               Say hello!
    -H                        Hello but in caps
    -t, --test                This is a test lol
    -f, --flag                Turns this flag on
    -o, --once                This flag may only be set once
    -i, --my-int-argument     [int] This argument expects an integer value
    -s, --my-string-argument  [string] This argument expects a string value
    --                        All subsequent arguments will not be processed as options
./full_example -f -i 42
--my-int-argument 42
All done with argument parsing!
./full_example -f --my-int-argument 42
--my-int-argument 42
All done with argument parsing!
./full_example -f --my-int-argument=42
--my-int-argument 42
All done with argument parsing!
./full_example -fi 42
--my-int-argument 42
All done with argument parsing!
./full_example -if 42
In argument "-if", option -i expects a value and therefore must be the last character.
./full_example -f test
ARG_POSITIONAL: test
Index: 2
./full_example -f -- here are some args
Raw argument after "--": "here"
Raw argument after "--": "are"
Raw argument after "--": "some"
Raw argument after "--": "args"
./full_example -tHfots lol --my-int-argument=42 -- test lol omg --help -o -t
Test: 1 2 3 4 5 6 7 8 9 10
HELLO
Test: 1 2 3 4 5 6 7 8 9 10
--my-string-argument lol
--my-int-argument 42
Raw argument after "--": "test"
Raw argument after "--": "lol"
Raw argument after "--": "omg"
Raw argument after "--": "--help"
Raw argument after "--": "-o"
Raw argument after "--": "-t"
./full_example -
ARG_POSITIONAL: -
Index: 1
./full_example ---
ARG_POSITIONAL: ---
Index: 1
./full_example -f- test
In argument "-f-", there is no supported option '-'
./full_example -=
ARG_POSITIONAL: -=
Index: 1
./full_example -=test
In argument "-=test", there is no supported option '='
./full_example --=
ARG_POSITIONAL: --=
Index: 1
./full_example --=test
ARG_POSITIONAL: --=test
Index: 1
